src/backend/distributed/planner/multi_physical_planner.c

```diff
@@ -1271,38 +1276,81 @@ JoinRangeTableEntry(JoinExpr *joinExpr, List *dependentJobList, List *rangeTable
 	rangeTableEntry->eref->colnames = joinedColumnNames;
 	rangeTableEntry->joinaliasvars = joinedColumnVars;
 
-	SetJoinRelatedColumnsCompat(rangeTableEntry,
-								leftColumnNames, rightColumnNames, leftColumnVars,
+	SetJoinRelatedColumnsCompat(rangeTableEntry, leftRelId, rightRelId, leftColumnVars,
 								rightColumnVars);
 
 	return rangeTableEntry;
 }
 
 
 static void
-SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry,
-							List *leftColumnNames, List *rightColumnNames,
+SetJoinRelatedColumnsCompat(RangeTblEntry *rangeTableEntry, Oid leftRelId, Oid rightRelId,
 							List *leftColumnVars, List *rightColumnVars)
 {
 	#if PG_VERSION_NUM >= PG_VERSION_13
 
 	/* We don't have any merged columns so set it to 0 */
 	rangeTableEntry->joinmergedcols = 0;
-	int numvars = list_length(leftColumnVars);
-	for (int varId = 1; varId <= numvars; varId++)
+
+	if (OidIsValid(leftRelId))
+	{
+		rangeTableEntry->joinleftcols = GetColumnOriginalIndexes(leftRelId);
+	}
+	else
+	{
+		int numvars = list_length(leftColumnVars);
+		for (int varId = 1; varId <= numvars; varId++)
+		{
+			rangeTableEntry->joinleftcols = lappend_int(rangeTableEntry->joinleftcols,
+														varId);
+		}
+	}
```

Comment:
This part might go into a function and then we can call it for left & right relations separately